README.md
Muffin‑OAuth
============

**Muffin‑OAuth** adds OAuth 1 and 2 support to the Muffin_ framework, enabling both client-side and server-side authentication flows.

.. image:: https://github.com/klen/muffin-oauth/workflows/tests/badge.svg
   :target: https://github.com/klen/muffin-oauth/actions
   :alt: Tests Status

.. image:: https://img.shields.io/pypi/v/muffin-oauth
   :target: https://pypi.org/project/muffin-oauth/
   :alt: PyPI Version

.. image:: https://img.shields.io/pypi/pyversions/muffin-oauth
   :target: https://pypi.org/project/muffin-oauth/
   :alt: Python Versions

.. contents::

Requirements
============

- Python >= 3.10
- Compatible with `asyncio`, `Trio`, and `Curio`

Installation
============

Install via pip:

    pip install muffin-oauth

Usage
=====

Here's a basic example using OAuth2:

```python
from muffin import Application
from muffin_oauth import OAuthPlugin

app = Application("auth-example")
oauth = OAuthPlugin()
oauth.setup(app, providers={
    "github": {
        "client_id": "...",
        "client_secret": "...",
        "authorize_url": "https://github.com/login/oauth/authorize",
        "access_token_url": "https://github.com/login/oauth/access_token",
        "api_base_url": "https://api.github.com"
    }
})

@app.route("/")
async def login(request):
    return await oauth.authorize_redirect(request, "github", redirect_uri="http://localhost:8000/callback")

@app.route("/callback")
async def callback(request):
    token = await oauth.authorize_access_token(request, "github")
    request.session["token"] = token
    return "Logged in"

@app.route("/user")
async def user(request):
    client = oauth.client("github", token=request.session.get("token"))
    resp = await client.get("/user")
    return resp.json()
```

Run the example app:

```bash
$ make example
http://localhost:5000
```

Client-side usage:

```python
client = oauth.client("github", access_token="...")
resp = await client.request("GET", "/user")
user_info = resp.json()
```

This supports both OAuth1 and OAuth2 flows, with automatic token handling and resource access via configured providers.

## Testing & Security

- Test coverage for major flows is provided in `tests.py`.
- Minimal dependencies and async-native design.
- Production-ready, MIT-licensed.

## Bug Tracker & Contributing

Found an issue or have an idea? Report it at:
https://github.com/klen/muffin-oauth/issues

Contributions welcome! Fork the repo and submit a PR.

## Contributors

- [klen](https://github.com/klen) (Kirill Klenov)

## License

Licensed under the [MIT license](http://opensource.org/licenses/MIT).

[Muffin]: https://github.com/klen/muffin


Changelog
2021-10-21  horneds

	* Support Python 3.10

2021-01-17  klen

	* Support Muffin > 0.40
	* Support Trio
	* Drop OAuth1 Support

2015-04-06  klen

	* Version 0.1.0
	* Initial release


tests.py
from unittest import mock

import muffin
import pytest
import sys
from yarl import URL


# XXX: python 3.7
def to_coro(value):
    async def coro():
        return value
    return coro()


@pytest.fixture
def app():
    import muffin_oauth

    app = muffin.Application(DEBUG=True, OAUTH_CLIENTS={
        'github': {
            'client_id': 'b6281b6fe88fa4c313e6',
            'client_secret': '21ff23d9f1cad775daee6a38d230e1ee05b04f7c',
            'params': {
                'scope': 'user:email'
            }
        }
    })

    oauth = muffin_oauth.Plugin(app)

    @app.route('/github')
    async def auth(request):
        client, token, data = await oauth.login('github', request)
        return await client.request('GET', '/user')

    return app


def test_client(app):
    from aioauth_client import GithubClient
    from muffin_oauth import OAuthException

    oauth = app.plugins['oauth']
    client = oauth.client('github')
    assert client
    assert isinstance(client, GithubClient)

    with pytest.raises(OAuthException):
        oauth.client('google')

    with pytest.raises(OAuthException):
        oauth.client('unknown')


async def test_muffin_oauth(app, client):
    res = await client.get('/github', follow_redirect=False)
    assert res.status_code == 307
    location = URL(res.headers['location'])
    assert location.host == 'github.com'
    state = location.query.get('state')
    assert state

    res = await client.get('/github?code=000')
    assert res.status_code == 406

    with mock.patch('aioauth_client.OAuth2Client._request') as mocked:
        mocked.return_value = {'access_token': 'test_passed'}
        if sys.version_info < (3, 8):
            mocked.side_effect = [
                to_coro(mocked.return_value),
                to_coro(mocked.return_value),
            ]

        res = await client.get('/github?code=000&state=%s' % state)
        assert res.status_code == 200
        assert await res.json() == {'access_token': 'test_passed'}


example.py
"""Example application.

Requirements:
    uvicorn

Run the example with uvicorn:

    $ uvicorn --port 5000 example:app

"""

import muffin
import muffin_oauth
import muffin_session
import html
from pprint import pformat


app = muffin.Application(
    'oauth', DEBUG=True,
    OAUTH_CLIENTS={
        'github': {
            'client_id': 'b6281b6fe88fa4c313e6',
            'client_secret': '21ff23d9f1cad775daee6a38d230e1ee05b04f7c',
            'scope': 'user:email',
        },
        'google': {
            'client_id': '150775235058-9fmas709maee5nn053knv1heov12sh4n.apps.googleusercontent.com',  # noqa
            'client_secret': 'df3JwpfRf8RIBz-9avNW8Gx7',
            'scope': 'profile email',
        },
    })

session = muffin_session.Plugin(app, secret_key='Example')
oauth = muffin_oauth.Plugin(app)


@app.route('/')
async def index(request):
    """Index Page."""
    return """
        <link rel="stylesheet"
            href="https://stackpath.bootstrapcdn.com/bootstrap/4.5.2/css/bootstrap.min.css" />
        <div class="container">
            <header class="navbar navbar-dark" style="background-color: #7952b3">
                <h2 class="navbar-brand">Muffin OAuth Example</h2>
            </header>
            <ul class="nav flex-column mt-5">
                <li class="nav-item">
                    <a class="nav-link" href="/oauth/github">Login with Github</a></li>
                <li class="nav-item">
                    <a class="nav-link" href="/oauth/google">Login with Google</a></li>
            </ul>
        </div>
    """


@app.route('/oauth/{provider}')
async def oauth_info(request):
    """Oauth example."""
    provider = request.path_params.get('provider')
    client, *_ = await oauth.login(provider, request)
    user, data = await client.user_info()
    return f"""
        <link rel="stylesheet"
            href="https://stackpath.bootstrapcdn.com/bootstrap/4.5.2/css/bootstrap.min.css" />
        <div class="container">
            <header class="navbar navbar-dark" style="background-color: #7952b3">
                <h2 class="navbar-brand">Muffin OAuth Example ({ client.name })</h2>
            </header>
            <a class="btn btn-primary mt-4" href='/'>Return back</a>
            <table class="table mt-4">
                <tr><td> ID </td><td> { user.id } </td></tr>
                <tr><td> Username </td><td> { user.username } </td></tr>
                <tr><td>First, last name</td><td>{ user.first_name }, { user.last_name }</td></tr>
                <tr><td>Gender</td><td> { user.gender } </td></tr>
                <tr><td>Email</td><td> { user.email } </td></tr>
                <tr><td>Link</td><td> { user.link } </td></tr>
                <tr><td>Picture</td><td> { user.picture } </td></tr>
                <tr><td>Country, City</td><td> { user.country }, { user.city } </td></tr>
            </table>
            <h3 class="mt-4">Raw data</h3>
            <pre>{ html.escape(pformat(data)) }</pre>
        </div>
    """


muffin_oauth.py
"""Support OAuth in Muffin Framework."""
from hashlib import sha1
from random import SystemRandom
import muffin
from aioauth_client import ClientRegistry
from muffin.plugin import BasePlugin


__version__ = "0.6.1"
__project__ = "muffin-oauth"
__author__ = "Kirill Klenov <horneds@gmail.com>"
__license__ = "MIT"

random = SystemRandom().random


class OAuthException(Exception):

    """Implement an exception during OAUTH process."""

    pass


class Plugin(BasePlugin):

    """Support OAuth."""

    name = 'oauth'
    defaults = {
        'clients': {},
        'redirect_uri': None,
    }

    def client(self, client_name, **params):
        """Initialize OAuth client from registry."""
        if client_name not in self.cfg.clients:
            raise OAuthException('Unconfigured client: %s' % client_name)

        if client_name not in ClientRegistry.clients:
            raise OAuthException('Unsupported services: %s' % client_name)

        params = dict(self.cfg.clients[client_name], **params)
        return ClientRegistry.clients[client_name](**params)

    async def authorize(self, client, session, redirect_uri=None, **params):
        """Get authorization URL."""
        state = sha1(str(random()).encode('ascii')).hexdigest()
        session['muffin_oauth'] = state
        return client.get_authorize_url(redirect_uri=redirect_uri, state=state, **params)

    async def login(self, client_name, request, redirect_uri=None, headers=None, **params):
        """Process login with OAuth.

        :param client_name: A name one of configured clients
        :param request: Web request
        :param redirect_uri: An URI for authorization redirect
        """
        client = self.client(client_name, logger=self.app.logger)
        session = self.app.plugins['session']

        redirect_uri = redirect_uri or self.cfg.redirect_uri or str(request.url)
        ses = session.load_from_request(request)

        code = request.query.get('code')
        if not code:
            url = await self.authorize(client, ses, redirect_uri, **params)
            res = muffin.ResponseRedirect(url)
            session.save_to_response(ses, res)
            raise res

        # Check state
        state = request.query.get('state')
        stored_state = ses.pop('muffin_oauth', '')
        if stored_state != state:
            raise muffin.ResponseError(406, 'Invalid state: "%s"' % stored_state)

        # Get access token
        return (
            client,
            await client.get_access_token(code, redirect_uri=redirect_uri, headers=headers)
        )

    def refresh(self, client_name, refresh_token, **params):
        """Get refresh token.

        :param client_name: A name one of configured clients
        :param redirect_uri: An URI for authorization redirect
        :returns: a coroutine
        """
        client = self.client(client_name, logger=self.app.logger)
        return client.get_access_token(refresh_token, grant_type='refresh_token', **params)


example/config.py
""" Setup the application. """

PLUGINS = 'muffin_session', 'muffin_oauth',
DEBUG = True

OAUTH_CLIENTS = {
    'github': {
        'client_id': 'b6281b6fe88fa4c313e6',
        'client_secret': '21ff23d9f1cad775daee6a38d230e1ee05b04f7c',
        'scope': 'user:email',
    },
    'bitbucket': {
        'consumer_key': '4DKzbyW8JSbnkFyRS5',
        'consumer_secret': 'AvzZhtvRJhrEJMsGAMsPEuHTRWdMPX9z',
    },
    'twitter': {
        'consumer_key': 'oUXo1M7q1rlsPXm4ER3dWnMt8',
        'consumer_secret': 'YWzEvXZJO9PI6f9w2FtwUJenMvy9SPLrHOvnNkVkc5LdYjKKup',
    },
    'facebook': {
        'client_id': '384739235070641',
        'client_secret': '8e3374a4e1e91a2bd5b830a46208c15a',
        'scope': 'email',
    },
    'google': {
        'client_id': '150775235058-9fmas709maee5nn053knv1heov12sh4n.apps.googleusercontent.com',
        'client_secret': 'df3JwpfRf8RIBz-9avNW8Gx7',
        'scope': 'profile email',
    },
    'yandex': {
        'client_id': 'e19388a76a824b3385f38beec67f98f1',
        'client_secret': '1d2e6fdcc23b45849def6a34b43ac2d8',
    }
}


muffin_oauth/__init__.py
"""Support OAuth in Muffin Framework."""
import base64
import hmac
import typing as t
from hashlib import sha1, sha512
from random import SystemRandom

from aioauth_client import ClientRegistry, Client
from muffin import Request, ResponseRedirect, ResponseError, Application
from muffin.plugins import BasePlugin


__version__ = "2.2.0"
__project__ = "muffin-oauth"
__author__ = "Kirill Klenov <horneds@gmail.com>"
__license__ = "MIT"

random = SystemRandom().random


class OAuthException(Exception):

    """Implement an exception during OAUTH process."""

    pass


class Plugin(BasePlugin):

    """Support OAuth."""

    name = 'oauth'
    defaults: t.Dict = {
        'clients': {},
        'redirect_uri': None,
        'secret': 'important:replace_me_asap!',
    }

    def client(self, client_name: str, **params) -> Client:
        """Initialize OAuth client from registry."""
        if client_name not in self.cfg.clients:
            raise OAuthException('Unconfigured client: %s' % client_name)

        if client_name not in ClientRegistry.clients:
            raise OAuthException('Unsupported services: %s' % client_name)

        params = dict(self.cfg.clients[client_name], **params)
        return ClientRegistry.clients[client_name](**params)

    def authorize(self, client: Client, redirect_uri: str = None, **params) -> str:
        """Get authorization URL."""
        state = sha1(str(random()).encode('ascii')).hexdigest()
        state = f"{ state }.{ sign(state, self.cfg.secret) }"
        return client.get_authorize_url(redirect_uri=redirect_uri, state=state, **params)

    async def login(
            self, client_name: str, request: Request, redirect_uri: str = None,
            headers: t.Dict = None, **params) -> t.Tuple[Client, str, t.Any]:
        """Process login with OAuth.

        :param client_name: A name one of configured clients
        :param request: Web request
        :param redirect_uri: An URI for authorization redirect
        """
        app = t.cast(Application, self.app)
        client = self.client(client_name, logger=app.logger)

        redirect_uri = redirect_uri or self.cfg.redirect_uri
        if not redirect_uri:
            redirect_uri = str(request.url.with_query(''))

        code = request.url.query.get('code')
        if not code:
            url = self.authorize(client, redirect_uri, **params)
            raise ResponseRedirect(url)

        # Check state
        state = request.url.query.get('state')
        if not state:
            raise ResponseError.NOT_ACCEPTABLE('Invalid state')

        state, _, sig = state.partition('.')
        if sig != sign(state, self.cfg.secret):
            raise ResponseError.NOT_ACCEPTABLE('Invalid state')

        # Get access token
        token, data = await client.get_access_token(
            code, redirect_uri=redirect_uri, headers=headers)
        return client, token, data

    async def refresh(self, client_name: str, refresh_token: str, **params) -> t.Tuple[str, t.Any]:
        """Get refresh token.

        :param client_name: A name one of configured clients
        :param redirect_uri: An URI for authorization redirect
        """
        app = t.cast(Application, self.app)
        client = self.client(client_name, logger=app.logger)
        return await client.get_access_token(refresh_token, grant_type='refresh_token', **params)


def sign(msg: str, key: str) -> str:
    """Sign the given message with the key."""
    sig = hmac.new(key.encode('utf-8'), msg.encode('utf-8'), sha512).digest()
    return base64.urlsafe_b64encode(sig).decode('utf-8')


example/app.py
"""Example application."""

import muffin
import html


app = muffin.Application(
    'oauth',
    PLUGINS=['muffin_session', 'muffin_oauth'],
    OAUTH_CLIENTS={
        'github': {
            'client_id': 'b6281b6fe88fa4c313e6',
            'client_secret': '21ff23d9f1cad775daee6a38d230e1ee05b04f7c',
            'scope': 'user:email',
        },
        'bitbucket': {
            'consumer_key': '4DKzbyW8JSbnkFyRS5',
            'consumer_secret': 'AvzZhtvRJhrEJMsGAMsPEuHTRWdMPX9z',
        },
        'twitter': {
            'consumer_key': 'oUXo1M7q1rlsPXm4ER3dWnMt8',
            'consumer_secret': 'YWzEvXZJO9PI6f9w2FtwUJenMvy9SPLrHOvnNkVkc5LdYjKKup',
        },
        'facebook': {
            'client_id': '384739235070641',
            'client_secret': '8e3374a4e1e91a2bd5b830a46208c15a',
            'scope': 'email',
        },
        'google': {
            'client_id': '150775235058-9fmas709maee5nn053knv1heov12sh4n.apps.googleusercontent.com',  # noqa
            'client_secret': 'df3JwpfRf8RIBz-9avNW8Gx7',
            'scope': 'profile email',
        },
        'yandex': {
            'client_id': 'e19388a76a824b3385f38beec67f98f1',
            'client_secret': '1d2e6fdcc23b45849def6a34b43ac2d8',
        }
    })


@app.register('/')
def index(request):
    """Index Page."""
    return """
        <ul>
            <li><a href="/oauth/bitbucket">Bitbucket</a></li>
            <li><a href="/oauth/facebook">Facebook</a></li>
            <li><a href="/oauth/github">Github</a></li>
            <li><a href="/oauth/google">Google</a></li>
            <li><a href="/oauth/twitter">Twitter</a></li>
            <li><a href="/oauth/yandex">Yandex</a></li>
        </ul>
    """


@app.register('/oauth/{provider}')
async def oauth(request):
    """Oauth example."""
    provider = request.match_info.get('provider')
    client, _ = await app.ps.oauth.login(provider, request)
    user, data = await client.user_info()
    response = (
        "<a href='/'>back</a><br/><br/>"
        "<ul>"
        "<li>ID: {u.id}</li>"
        "<li>Username: {u.username}</li>"
        "<li>First, last name: {u.first_name}, {u.last_name}</li>"
        "<li>Email: {u.email}</li>"
        "<li>Link: {u.link}</li>"
        "<li>Picture: {u.picture}</li>"
        "<li>Country, city: {u.country}, {u.city}</li>"
        "</ul>"
    ).format(u=user)
    response += "<code>%s</code>" % html.escape(repr(data))
    return response
